spanconfig: support rangefeeds for dynamic system tables

irfansharif opened the pull request:
Fixes #73045.

We're running out of system table IDs (see #57531), and as a result
we're now introducing the notion of dynamic system IDs throughout the
system. Previously KV baked-in the assumption of static system IDs at
two points:
- When deciding to allow rangefeeds on a given range;
- When enforcing strict GC TTL;

It did so by decoding the range's key span and comparing against the
hard-coded maximum system ID, all to determine whether the range in
question contained system tables. If so, we allowed rangefeeds to be
declared over it, and also did not enforce strict GC TTL (only really
applies to user tables). This way of doing things does not compose with
dynamically allocated system table IDs. With arbitrary, possibly
non-contiguous IDs, we don't have the convenient key-comparison
properties to rely on.

To that end, we use the span configs infrastructure to to delegate
control of whether rangefeeds are enabled over a given range and whether
strict GC is enforced. This scheme allows SQL code to declare "system
table configs" over arbitrary schemas, and have KV still respect it.
This PR does not expose these span config settings as part of zone
configs -- there's no need to (though we could in the future). To
account for the asynchronous nature of the span configs infra, we need
to ensure that ranges without an available config default to enabling
rangefeeds.

Release note: None

irfansharif committed:
spanconfig: support rangefeeds for dynamic system tables

Fixes #73045.

We're running out of system table IDs (see #57531), and as a result
we're now introducing the notion of dynamic system IDs throughout the
system. Previously KV baked-in the assumption of static system IDs at
two points:
- When deciding to allow rangefeeds on a given range;
- When enforcing strict GC TTL;

It did so by decoding the range's key span and comparing against the
hard-coded maximum system ID, all to determine whether the range in
question contained system tables. If so, we allowed rangefeeds to be
declared over it, and also did not enforce strict GC TTL (only really
applies to user tables). This way of doing things does not compose with
dynamically allocated system table IDs. With arbitrary, possibly
non-contiguous IDs, we don't have the convenient key-comparison
properties to rely on.

To that end, we use the span configs infrastructure to to delegate
control of whether rangefeeds are enabled over a given range and whether
strict GC is enforced. This scheme allows SQL code to declare "system
table configs" over arbitrary schemas, and have KV still respect it.
This PR does not expose these span config settings as part of zone
configs -- there's no need to (though we could in the future). To
account for the asynchronous nature of the span configs infra, we need
to ensure that ranges without an available config default to enabling
rangefeeds.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74555)
<!-- Reviewable:end -->

irfansharif commented:
This is "done", but still marking it as draft; it's building on top of 3 open PRs that I'd like to land first.

postamar commented on pkg/spanconfig/spanconfigsqltranslator/sqltranslator.go:
```diff
@@ -230,9 +229,18 @@ func (s *SQLTranslator) generateSpanConfigurationsForTable(
 		return nil, err
 	}
 
+	isSystemID := s.execCfg.SystemIDChecker.IsSystemID(uint32(desc.GetID()))
```
FYI, since you already have the descriptor you can call `catalog.IsSystemDescriptor(desc)` instead.

irfansharif replied:
Done.

ajwerner reviewed (approved):
Code <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, interesting that the tests are failing in the way that they are.  Any ideas what's going on there?

> Reviewed 13 of 92 files at r1, 79 of 79 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74555#-:-MtyIIN31O-ImWKmvAsx:b1wbjnd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif)



<!-- Sent from Reviewable.io -->

irfansharif commented:
Not yet, but thankfully it's contained only to this PR. It'll take some time to munch through, and I'm a bit occupied with the fallout from https://github.com/cockroachdb/cockroach/pull/73876. I'll let this PR sit on ice for a bit.

irfansharif commented:
TFTR!

>  interesting that the tests are failing in the way that they are. Any ideas what's going on there?

Just a few tests that needed updating.

bors r+

irfansharif committed:
kvserver: de-flake TestRangefeedWorksOnSystemRangesUnconditionally

Now that we're using span configs to control whether rangefeeds are
enabled on ranges or not, this test need some additional synchrony
(waiting for span configs to be applied).

Release note: None

irfansharif commented:
bors r+

ajwerner commented:
This most recent flake looks plausibly legit. 

```
=== RUN   TestProposalOverhead/user-key_overhead
=== CONT  TestProposalOverhead/user-key_overhead
    client_replica_test.go:3603:
          Error Trace:  client_replica_test.go:3603
          Error:        Not equal:
                        expected: 0x2a
                        actual  : 0x45
          Test:         TestProposalOverhead/user-key_overhead
```
https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests/4256737?showRootCauses=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true

I imagine this could happen if the logical op log was still enabled.

bors r-

irfansharif commented:
Ugh, thanks for flagging. I keep catching new flakes every CI run.

irfansharif committed:
kvserver: de-flake TestProposalOverhead and ..

..TestReplicaRangefeedRetryErrors.

- TestProposalOverhead relies on configs applying to the scratch range
  now that we're using configs to determine whether rangefeeds are
  enabled or not. We just opt into the right testing knob.
- TestReplicaRangefeedRetryErrors creates a split from "underneath" the
  span configs infra, which previously ended up applying the default
  config (rangefeed-enabled=true) for the freshly created replicas. This
  is not what the test wants. Short of rewriting the test, we introduce
  an interceptor to manually override the replica's span config.

Release note: None

irfansharif commented:
> This most recent flake looks plausibly legit.

Just more tests that needed updating (in appended commits). Will merge on green.

irfansharif commented:
bors r+

yuzefovich commented:
Tracked down one of the regressions in microbenchmark to the first commit here:
```
Switching to 4b8258c4b4a8f89fe8a4ea87fc8ee7cd5c916b6d^
+ make bench PKG=./pkg/bench BENCHTIMEOUT=24h BENCHES=BenchmarkSQL/MultinodeCockroach/InsertSecondaryIndex/count=1000 'TESTFLAGS=-count 10 -benchmem'
Switching to 4b8258c4b4a8f89fe8a4ea87fc8ee7cd5c916b6d
+ make bench PKG=./pkg/bench BENCHTIMEOUT=24h BENCHES=BenchmarkSQL/MultinodeCockroach/InsertSecondaryIndex/count=1000 'TESTFLAGS=-count 10 -benchmem'
name                                                       old time/op    new time/op    delta
SQL/MultinodeCockroach/InsertSecondaryIndex/count=1000-24    59.9ms ±12%    64.1ms ±15%     ~     (p=0.123 n=10+10)

name                                                       old alloc/op   new alloc/op   delta
SQL/MultinodeCockroach/InsertSecondaryIndex/count=1000-24     198MB ±31%     249MB ±30%  +25.77%  (p=0.011 n=10+10)

name                                                       old allocs/op  new allocs/op  delta
SQL/MultinodeCockroach/InsertSecondaryIndex/count=1000-24      394k ± 3%      562k ± 3%  +42.84%  (p=0.000 n=10+10)
```
Not sure if we can do anything about it.

irfansharif commented:
I looked into this a bit, no smoking gun other than it being clearly due to more rangefeed processing. Filed #79602 to track investigation. (+cc @ajwerner just in case the additional mem usage looks patently obvious to you.)
